### ISSUE TITLE
DDP NCCL Parameters For Performance

### ISSUE THREAD

**Issue by shuyingsunshine21** [PyTorchLightning/pytorch-lightning#7179]
### Motivation

From several experiments, DDP on NCCL backend 
* NCCL_NSOCKS_PERTHREAD (https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-nsocks-perthread)
* NCCL_SOCKET_NTHREADS (https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-socket-nthreads)
* NCCL_MIN_NCHANNELS (https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-min-nchannels)
are important environment parameters to tune for optimizing communication performance.

For example: XLM-RoBERTa (https://arxiv.org/abs/1911.02116), 30% speedup for `NCCL_NSOCKS_PERTHREA = 4` and `NCCL_SOCKET_NTHREADS = 2`

Detectron2 (https://github.com/facebookresearch/detectron2), 15% speedup for `NCCL_NSOCKS_PERTHREA = 4` and `NCCL_SOCKET_NTHREADS = 2`

### Pitch

we could pass these parameters from  `kwargs`  (similarly as `find_unused_parameters`), and set these parameters in `setup_environment` when initializing ddp process.

**Comment by justusschock** [comment]
Hi @shuyingsunshine21 
I am not that familiar with NCCL; Do you see any downside to use these defaults?

**Comment by shuyingsunshine21** [comment]
> Hi @shuyingsunshine21
> I am not that familiar with NCCL; Do you see any downside to use these defaults?

let me rephrase the pitch a bit, maybe not setting default, instead relying on users to pass in from `kwargs` for `DDPPlugin`

**Comment by SciPioneer** [comment]
We have seen good speedups by tuning these parameters on different domains (NLP, speech, vision, etc), as long as the training has a large communication overheads.

**Comment by ananthsub** [comment]
@SciPioneer are these dependent on the cluster setup? for example, would we expect AWS or GCP to set these parameters on their own? is their risk of lightning the framework overriding these values?

**Comment by SciPioneer** [comment]
@ananthsub Both cluster admin and individual users can set their own default values, in `/etc/nccl.conf` and `~/.nccl.conf`, respectively. Actually both AWS and GCP have set their own default values. This does not hold true for every cluster though.

I think this proposal aims to allow the users to specify a set of application-specific default values. As long as lightning doesn't override the default values, there shouldn't be any risk.

**Comment by shuyingsunshine21** [comment]
this is not limited to DDP plugin, setting these when initiating `ClusterEnvironment` makes sense, see PR #7794

**Comment by edenlightning** [comment]
Decided to add these parameters to the speed documentation instead: https://pytorch-lightning.readthedocs.io/en/latest/guides/speed.html#gpu-training
